mitiq/lre/inference/multivariate_richardson.py

```diff
@@ -0,0 +1,156 @@
+            extrapolation upto max degree
+    """
+    variable_combinations = _create_variable_combinations(num_layers, degree)
+
+    monomial_basis = []
+    for combo in variable_combinations:
+        monomial_parts = []
+        counts = Counter(combo)
+        # Ensure variables are processed in lexicographical order
+        for var in sorted(counts.keys()):
+            count = counts[var]
+            if count > 1:
+                monomial_parts.append(f"{var}**{count}")
+            else:
+                monomial_parts.append(var)
+        monomial = "*".join(monomial_parts)
+        # Handle the case where degree is 0
+        # the tuple in variable_combinations is empty
+        if not monomial:
+            monomial = "1"
+        monomial_basis.append(monomial)
+    # "1" should be the first monomial (degree = 0).
+    # max degree should be the last term
+    return monomial_basis[::-1]
+
+
+def sample_matrix(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> NDArray[Any]:
+    """Calculates the sample matrix required for extrapolation."""
+    if degree < 1:
+        raise ValueError(
+            "Multinomial degree must be greater than or equal to 1."
+        )
+    if fold_multiplier < 1:
+        raise ValueError("Fold multiplier must be greater than or equal to 1.")
+
+    num_layers = _get_num_layers_without_measurements(input_circuit)
+
+    scale_factor_vectors = _get_scale_factor_vectors(
+        input_circuit, degree, fold_multiplier, num_chunks
+    )
+
+    monomial_terms = full_monomial_basis(num_layers, degree)
+    if len(monomial_terms) != len(scale_factor_vectors):
+        raise ValueError("Sample matrix will not be a square matrix.")
+    sample_matrix = np.zeros((len(monomial_terms), len(monomial_terms)))
+
+    # Evaluate the monomial terms using the values in the scale factor vectors
+    # and insert in the sample matrix
+    # each row is specific to each scale factor vector
+    # each column is a term in the monomial basis
+    for i, point in enumerate(scale_factor_vectors):
+        for j, monomial in enumerate(monomial_terms):
+            var_mapping = {f"λ_{k+1}": point[k] for k in range(num_layers)}
+            sample_matrix[i, j] = eval(monomial, {}, var_mapping)
+    return sample_matrix
+
+
+def linear_combination_coefficients(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> List[int]:
+    """Finds the coefficients according to equation 20."""
+    num_layers = len(
+        _get_scale_factor_vectors(
+            input_circuit, degree, fold_multiplier, num_chunks
+        )
+    )
+    input_sample_matrix = sample_matrix(
+        input_circuit, degree, fold_multiplier, num_chunks
+    )
+
+    if np.linalg.det(input_sample_matrix) == 0:
+        raise ValueError("Sample matrix is singular.")
```

Comment:
It might be better to remove this check. 

If the determinant is zero, we should get the division by zero error.  

![image](https://github.com/user-attachments/assets/3799e0b0-f818-458a-bcc2-be7b42f0a954)